test-script/gen_commands.py

```diff
@@ -0,0 +1,34 @@
+import glob
+import os
+import argparse
+
+modes = ['atomicity-violation', 'observability-gap', 'time-travel']
+mode_map = {
+    'atomicity-violation': 'atom-vio',
+    'observability-gap': 'obs-gap',
+    'time-travel': 'time-travel'
+}
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Generate testcase commands into a file.")
+    parser.add_argument('-d', dest='docker', help="Docker account", default='ghcr.io/sieve-project/action')
+    parser.add_argument('-o', dest='output', help="Output file name", default='commands.txt')
+    args = parser.parse_args()
+
+    with open(args.output, 'w') as command_file, open('pull-commands.txt', 'w') as pull_command_file:
+        # pull all k8s images
+        pull_command_file.write('docker pull {}/node:{}\n'.format(args.docker, 'learn'))
+        pull_command_file.write('docker pull {}/node:{}\n'.format(args.docker, 'time-travel'))
+        pull_command_file.write('docker pull {}/node:{}\n'.format(args.docker, 'observability-gap'))
+        pull_command_file.write('docker pull {}/node:{}\n'.format(args.docker, 'atomicity-violation'))
+
+        for operator in os.listdir('../log'):
```

Comment:
gen_commands.py is only ran once on the master node, then it generates all the testing command with specific config. So basically we will have a pool of test commands which gets distributed to each node. The GNU parallel accepts a list of commands and a list of hosts, it will run commands distributively on the list of nodes.

Reply:
Oh I see. This is achieved by `-j 1`, right? This is neat.

Reply:
The `-j 1` command means that we only want one test to be running on each node at a time. GNU parallel by its nature just run each command only once, regardless on which machine. Note that I specifies `--onall` flag in step 2, which requires GNU parallel to run the same list of commands on all machines. But without this `--onall` flag, each command is ran only once.